app/api/sul_bib/publications_api.rb

```diff
@@ -17,37 +16,18 @@ def publication_find(id)
         error!({ 'error' => 'No such publication', 'detail' => "You've requested a non-existant publication." }, 404)
       end
 
-      # Check for existing authors.
-      # @param authorship_list [Array<Hash>]
-      # @return [boolean] true if any of the authors exist.
-      def existing_authors?(authorship_list)
-        # At least one of the authors in the authorship array must exist.
-        return false if authorship_list.nil?
-        authorship_list.any? { |auth| Contribution.authorship_valid?(auth) }
-      end
-
       # Check for existing authors or create new authors with a CAP profile ID.
+      # At least one of the authors in the authorship array must exist or have
+      # a CAP profile that is used to create a new SULCAP author.
       # @param authorship_list [Array<Hash>]
-      # @return [boolean] true if any authors exist or are created.
+      # @return [Boolean] true if any authors exist or are created.
       def validate_or_create_authors(authorship_list)
-        # At least one of the authors in the authorship array must exist or have
-        # a CAP profile that is used to create a new SULCAP author.
-        return false if authorship_list.nil?
-        status = false
-        authorship_list.each do |authorship|
-          # Contribution.authorship_valid? confirms existing authors;
-          # it does not create new authors.
-          if Contribution.authorship_valid?(authorship)
-            status = true
-          else
-            cap_profile_id = authorship[:cap_profile_id]
-            unless cap_profile_id.blank?
-              Author.fetch_from_cap_and_create(cap_profile_id)
-              status = true
-            end
-          end
-        end
-        status
+        return false if authorship_list.blank?
+        groups = authorship_list.group_by { |auth| Contribution.authorship_valid?(auth) }
+        unknowns = groups[false] || []
+        cap_ids = unknowns.map { |auth| auth[:cap_profile_id] }.compact
+        cap_ids.map { |id| Author.fetch_from_cap_and_create(id) }.any? ||
+          groups[true].any?
       end
```

Comment:
Generally OK with this, if I understand it correctly, it behaves the same way as before.  It might introduce extra loops, but it's not important here because the authorship array is not large.

The only significant concern is that `Author.fetch_from_cap_and_create(id)` could throw an exception somewhere that could blow up before it gets to the boolean `.any?` calls.  (The prior code sorta had this problem too, it seems; so the code enforces that `cap_profile_id` params must belong to real records somewhere, but the API might return a 500 instead of a more graceful handling of it.)  Using the `Author.fetch_from_cap_and_create(id)` is an attempt to fulfill the request when we don't have it already and it can be retrieved; I'm not entirely sure what the behavior should be when it can't be retrieved and/or saved.  It does kinda significantly complicate the API because it introduces a "loop back" call to the API that the request is coming from in the first place; kinda prompts the thought about whether the response should throw the problem back to the client, which could then hit a different API endpoint to first add a new author resource (but that's not the responsibility of this PR, of course).

Reply:
Yeah, the design of this original method seems off.  To me, it seems like it really only matters that the **first** valid author is found.  But the existing code goes on to try to fetch/create any other authors also, without any provision for that possibly failing.  Certainly the execution behavior would be more predictable if at most 1 author were created, but I didn't want to change that here.